### ISSUE TITLE
Address Specifying Uniform Interface Method for protocol

### ISSUE THREAD

**Issue by fosrias** [ionwg/ion-doc#4]
Self-explanatory.

**Comment by lhazlewood** [comment]
Checking back in on this - I think this makes a lot of sense, but it sounds like it might be better suited to an additional Ion spec.  In other words, I'd guess that the 'base' spec defines the general Ion rules that apply no matter the communication protocol.

Another spec doc would build on that and specify specific fields/structure for HTTP.

In other words, there is the data resource format itself.  Then there is how to exchange resources over a given protocol.  So far nothing in Ion forces HTTP (yet) - even 'href' just means `hypermedia reference` which is an IRI and not necessarily a URL.

Take a look at SCIM (http://www.simplecloud.info/) and how they structured the SCIM docs.  Makes sense to follow that same parallel here to me.

Thoughts?

**Comment by fosrias** [comment]
Hm. Absent defining a method in the base spec then it just becomes a data schema spec vs. a hypermedia media type spec. Seems to me that including a `method` and some `enctype` information in the base spec makes it a hypermedia type out of the box. A further spec for HTTP could expand on that. 

The other option is to go the [uber](http://rawgit.com/uber-hypermedia/specification/master/uber-hypermedia.html) route where you define a method type and then define what that is in different protocols, albeit I saw some problems with this in that it was too opinionated.

So, is ion a data schema spec or a hypermedia media-type, IYO @lhazlewood?

**Comment by lhazlewood** [comment]
@fosrias I think just by having links makes it a valid hypermedia type.

Now how GET, PUT, POST and DELETE should be used to interact with collections, search, and CRUD for individual instances seems to be worthy of an additional spec document IMO.  In other words, if I'm building a REST API from scratch, I shouldn't have to figure *any* of this stuff myself and would like to look to a doc.  However, this spec shouldn't necessarily force concerns on people not building REST APIs.  

REST APIs were my initial motivation for Ion, but as we work on it, it seems like just getting the foundational document format + links seems to be core priority and then we can build on top of it for codifying REST APIs.

For example, if you add `method` to what we currently have, wouldn't we also need to indicate what happens, say, for `POST`: how to list what data fields are available vs required, what validation rules exist for each, any placeholder text, etc.  Basically replicating HTML's <input> concept as JSON objects (which I'm totally fine with and feel it is absolutely necessary for protocol interaction).

One thing to keep in mind is that Ion as it currently stands (and even with its `meta` `href` field, i.e. `hypermedia reference`) does not mandate or assume HTTP.  IRIs can be used in other protocols as well.  Do we want to force HTTP semantics?  My gut feeling is that we probably shouldn't do that and instead have an 'HTTP' profile on top of the core.

In other words, I can see a few specs (just like Atom, SCIM, HTTP and others):

* Ion
* Ion HTTP (builds on top of Ion core)
* Ion REST (builds on top of HTTP and specifically addresses REST concerns).

This layering approach feels safer and more flexible and satisfies many use case - plus you only choose the one you care about.

I should also note that, at least in my head, this doesn't require separate media types: the Meta Object Field Registry can contain any/all additional field names that would be relevant for HTTP.

Thoughts?

**Comment by fosrias** [comment]
@lhazlewood Chewing on this one a bit still, BTW.

**Comment by fosrias** [comment]
@lhazlewood @smizell  One thought rolling around is that absence defining the whole spectrum of controls in Ion (that is links and forms), there is no significant difference functionally between HAL and Ion. Basically, it is just the semantics of the media-type but by and large there is no significant advancement functionally. 

This is a major gap in HAL in many's opinion. Just some food for thought about specifying at least the form like controls around a specified uniform interface that is REST based and not protocol based. Then I think the Protocol specific aspects can be much lighter.

My two bits to no just be in a protocol spec per se. Frankly, it might be more useful to put in one to start and pull it out later. Not sure if atom protocol would have be done if it was not an after thought when atom spec was done.

**Comment by lhazlewood** [comment]
@fosrias and @smizell I added a `method` meta member in the member registry that is a single string value that is expected to match one of those in the HTTP specification (i.e. one method per transition as indicated above).  With the form stuff in #3, I'm pretty confident that my comments above will be adequately addressed.

I also have changed my mind about multiple specs.  Just the one should be satisfactory and is now more of a proper hypermedia type, rather than just a data format as @fosrias indicate.  Additionally, the form stuff will further distinguish Ion as a better format for REST than HAL (IMO).

As a result, I'm closing this issue, but please feel free to re-open if you think the HTTP methods aren't sufficient, or should be clarified in a different way.  Thanks for your continued feedback/support!
